[AMDGPU] Add no return image_sample intrinsics and instructions

perlfu opened the pull request:
An appropriately configured image resource descriptor can trigger image_sample instructions to store outputs directly to a linked memory location instead of returning to VGPRs.

This is opaque to the backend as instruction encoding is unchanged; however, a mechanism is require to allow frontends to communicate that these instructions do not require destination VGPRs and store to memory.
Flagging these as stores means they will not be optimized away.

perlfu committed:
[AMDGPU] Add no return image_sample intrinsics and instructions

An appropriately configured image resource descriptor can trigger
image_sample instructions to store outputs directly to a linked
memory location instead of returning to VGPRs.

This is opaque to the backend as instruction encoding is unchanged;
however, a mechanism is require to allow frontends to communicate
that these instructions do not require destination VGPRs and store
to memory.
Flagging these as stores means they will not be optimized away.

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -970,6 +984,21 @@ defset list<AMDGPUImageDimIntrinsic> AMDGPUImageDimIntrinsics = {
+  multiclass AMDGPUImageDimSampleNoReturnDims<string opmod,
+                                      AMDGPUSampleVariant sample> {
+    foreach dim = AMDGPUDims.NoMsaa in {
+      def !strconcat(NAME, "_", dim.Name) : AMDGPUImageDimIntrinsic<
```
If it's the same signature as the existing cases, can you just overload the return type to void?

perlfu replied:
The attributes are different, so I don't think they can share signatures?

arsenm replied:
You could rely on callsite attributes but that's a big hassle

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -780,6 +780,20 @@ class AMDGPUDimSampleProfile<string opmod,
+  let IsSample = true;
+  let RetTypes = [];
+  let DataArgs = [];
+  let ExtraAddrArgs = sample.ExtraAddrArgs;
+  let Offset = sample.Offset;
+  let Bias = sample.Bias;
+  let ZCompare = sample.ZCompare;
+  let Gradients = sample.Gradients;
+  let LodClampMip = sample.LodOrClamp;
```
This copies everything except the return?

perlfu replied:
Done

arsenm commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.sample.noret.ll:
```diff
@@ -0,0 +1,479 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1030 -verify-machineinstrs -global-isel=0 < %s | FileCheck -check-prefixes=GFX10PLUS,GFX10PLUS-SDAG,GFX10,GFX10-SDAG %s
```
Drop all the -verify-machineinstrs, and move the -global-isel arguments to be first

perlfu replied:
Done

arsenm commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.sample.noret.ll:
```diff
@@ -0,0 +1,479 @@
+; GFX12-NEXT:    image_sample_b off, [v0, v1], s[0:7], s[8:11] dmask:0xf dim:SQ_RSRC_IMG_1D
+; GFX12-NEXT:    s_endpgm
+main_body:
+  call void @llvm.amdgcn.image.sample.noret.b.1d.f32(i32 15, float %zcompare, float %s, <8 x i32> %rsrc, <4 x i32> %samp, i1 0, i32 0, i32 0)
```
Can we get the "noret" to be a pure suffix, so that the prefix of the name is common with the non-noret intrinsic? 

Also, I believe the manual calls this "rtn" vs "nortn"

perlfu replied:
I might be misunderstanding, but I think you are asking why this cannot be `llvm.amdgcn.image.sample.b.1d.f32.noret` ?
We need to select an entirely different intrinsic to get alternate attributes (operation has side-effects and cannot be optimized).
So I don't think a pure suffix works for that?

As for naming, yes I think maybe this makes more sense as "nortn", as we already use "rtn" for various instructions.

arsenm replied:
I mean intrinsic names should match the underlying instruction name as closely as possible. Here, you've split the opcode name with the compiler invented infix of noret. It would be better if this were llvm.amdgcn.image.sample.b.1d.nortn.f32. That way it always sorts with llvm.amdgcn.image.sample.b.1d.f32

perlfu replied:
Done

perlfu committed:
Address reviewer feedback:
- Rename _noret to _nortn
- Move intrinsic suffix to after dimensions to match instructions
- Deduplicate AMDGPUDimSampleNoReturnProfile definition
- Adjust test RUN lines

perlfu commented:
Gentle ping

perlfu commented:
Ping
